src/vs/workbench/contrib/notebook/browser/notebookEditor.ts

```diff
@@ -226,7 +229,26 @@ export class NotebookEditor extends EditorPane implements INotebookEditorPane {
 			}
 
 			if (model === null) {
-				throw new Error(localize('fail.noEditor', "Cannot open resource with notebook editor type '{0}', please check if you have the right extension installed and enabled.", input.viewType));
+				const knownProvider = this._notebookService.getViewTypeProvider(input.viewType);
+
+				if (knownProvider) {
+					throw createEditorOpenError(new Error(localize('fail.noEditor', "Cannot open resource with notebook editor type '{0}', please check if you have the right extension installed and enabled.", input.viewType)), [
+						toAction({
+							id: 'workbench.notebook.action.installMissing', label: localize('notebookOpenInstallMissingViewType', "Install extension for '{0}'", input.viewType), run: async () => {
+								const d = this._notebookService.onAddViewType(viewType => {
+									if (viewType === input.viewType) {
```

Comment:
I wonder if this is a bit dangerous leaving the listener around even when the `InstallRecommendedExtensionAction` is failing. I would throw in another dispose to clean up in case the action is not executing properly.